pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,531 @@
+	"github.com/cilium/ebpf/asm"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/ip"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+// DevicesControllerCell registers a controller that subscribes to network devices
+// and routes via netlink and populates the devices and routes devices.
+var DevicesControllerCell = cell.Module(
+	"datapath-devices-controller",
+	"Populates the device and route tables from netlink",
+
+	cell.Invoke(registerDevicesController),
+
+	// DeviceManager implements the old API for detecting and watching
+	// devices.
+	cell.Provide(newDeviceManager),
+)
+
+var (
+	// batchingDuration is the amount of time to wait for more
+	// addr/route/link updates before processing the batch.
+	batchingDuration = 100 * time.Millisecond
+
+	excludedDevicePrefixes = []string{
+		defaults.HostDevice,
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+		"sit",
+	}
+
+	// Route filter to look at all routing tables.
+	routeFilter = netlink.Route{
+		Table: unix.RT_TABLE_UNSPEC,
+	}
+	routeFilterMask = netlink.RT_FILTER_TABLE
+)
+
+type DevicesConfig struct {
+	Devices []string
+}
+
+type devicesControllerParams struct {
+	cell.In
+
+	Log         logrus.FieldLogger
+	DB          statedb.DB
+	DeviceTable statedb.Table[*tables.Device]
+	RouteTable  statedb.Table[*tables.Route]
+	Config      DevicesConfig
+
+	// NetNS is an optional network namespace handle. Used by tests to restrict the detection
+	// to specific namespace.
+	NetNS *netns.NsHandle `optional:"true"`
+}
+
+func registerDevicesController(lc hive.Lifecycle, p devicesControllerParams) {
+	dc := &devicesController{
+		devicesControllerParams: p,
+		links:                   map[int]netlink.Link{},
+		filter:                  deviceFilter(p.Config.Devices),
+	}
+	dc.ctx, dc.cancel = context.WithCancel(context.Background())
+	lc.Append(dc)
+}
+
+type devicesController struct {
+	devicesControllerParams
+	filter         deviceFilter
+	l3DevSupported bool
+	handle         *netlink.Handle
+	ctx            context.Context
+	cancel         context.CancelFunc
+	links          map[int]netlink.Link
+}
+
+func (dc *devicesController) Start(hive.HookContext) error {
+	dc.l3DevSupported = probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
+
+	// Construct a handle that uses the configured network namespace.
+	netns := netns.None()
+	if dc.NetNS != nil {
+		netns = *dc.NetNS
+	}
+	var err error
+	dc.handle, err = netlink.NewHandleAt(netns)
+	if err != nil {
+		return fmt.Errorf("NewHandleAt failed: %w", err)
+	}
+
+	// Subscribe to address updates to find out about link address changes
+	// to update the device's addresses.
+	addrUpdates := make(chan netlink.AddrUpdate, 16)
```

Comment:
It's just a small amount of buffering to avoid unnecessary switches between the goroutine doing `recvfrom` the netlink socket and this one batching them up. The actual amount doesn't matter much. This is now lifted into a constant defined at the top.